### PR TITLE
[Tagger] Allow collectors to skip caching tags on error

### LINKED ISSUE #6758: DD Agent does not retry resource tags collection for containers on ECS
Hi,

We have a problem with Datadog StatsD metrics missing tags when a new ECS task or instance is started. This happens randomly with less than 1% of metrics. We run our services in containers in AWS ECS, with each Container Instance (i.e. EC2 instance) having 1 datadog agent as a daemon task. Containers send StatsD metrics through a socket to the datadog agent. The datadog agent is configured to collect AWS tags of the tasks and EC2 instances (`DD_ECS_COLLECT_RESOURCE_TAGS_EC2`).
it looks like when datadog agent starts it sometimes doesn't collect tags of AWS resources such as EC2 instances and ECS tasks. This happens rarely (<1% of the time) and unpredictably and persists until the agent is restarted.

There is a corresponding Datadog support ticket (#402133). I'm creating this GH issue to provide some more technical details. I've done a lot of experiments and checked the source code of the datadog agent and here is the explanation of what is going wrong and why we lose tags.

## Tags caching

First, I would like to clarify the Datadog agent behaviour. When a new container starts on the host the agent tries to get tags for this container entity. If it succeeds the tags for this container are cached in memory and never fetched again. If it fails it will retry to fetch tags later until it succeeds. Here is the corresponding source code: [https://github.com/DataDog/datadog-agent/blob/666c775b3b0b408bc8ba4593f420fd526cd841a1/pkg/tagger/tagger.go#L234-L280](https://github.com/DataDog/datadog-agent/blob/666c775b3b0b408bc8ba4593f420fd526cd841a1/pkg/tagger/tagger.go#L234-L280)
The caching part is crucial here, please correct me if I'm wrong.

## How ECS tags get fetched

Let's dig deep into how the ECS tags get fetched for a single container entity. The entry point is `ECSCollector.Fetch` method: [https://github.com/DataDog/datadog-agent/blob/666c775b3b0b408bc8ba4593f420fd526cd841a1/pkg/tagger/collectors/ecs_main.go#L75](https://github.com/DataDog/datadog-agent/blob/666c775b3b0b408bc8ba4593f420fd526cd841a1/pkg/tagger/collectors/ecs_main.go#L75)
What it does:

1. Calls the `/v1/tasks` endpoint on the ECS agent running on the underlying host to get a list of tasks and their containers running on the host.
2. Checks whether the setting to collect resource tags `ecs_collect_resource_tags_ec2` is enabled and the metadata endpoint in ECS exposes resource tags: [https://github.com/DataDog/datadog-agent/blob/666c775b3b0b408bc8ba4593f420fd526cd841a1/pkg/tagger/collectors/ecs_main.go#L88](https://github.com/DataDog/datadog-agent/blob/666c775b3b0b408bc8ba4593f420fd526cd841a1/pkg/tagger/collectors/ecs_main.go#L88)
3. Parses the response got in 1. and looks up the task and the target container. Creates a list of tags from this data, e.g. `task_name`, `ecs_container_name`, etc. [https://github.com/DataDog/datadog-agent/blob/666c775b3b0b408bc8ba4593f420fd526cd841a1/pkg/tagger/collectors/ecs_extract.go#L19](https://github.com/DataDog/datadog-agent/blob/666c775b3b0b408bc8ba4593f420fd526cd841a1/pkg/tagger/collectors/ecs_extract.go#L19)
4. If the collection of resource tags is configured and supported, calls additional handler `addTagsForContainer`: [https://github.com/DataDog/datadog-agent/blob/666c775b3b0b408bc8ba4593f420fd526cd841a1/pkg/tagger/collectors/ecs_main.go#L116-L124](https://github.com/DataDog/datadog-agent/blob/666c775b3b0b408bc8ba4593f420fd526cd841a1/pkg/tagger/collectors/ecs_main.go#L116-L124) to collect tags set on the ECS task and instance using the `/v3/<uuid>/taskWithTags` endpoint on the ECS agent. Then adds the received tags to the resulting list of tags.
5. If `addTagsForContainer` in 4. fails to get tags from ECS metadata endpoint, only logs a warning "Unable to get resource tags for container..." but does not return any errors: [https://github.com/DataDog/datadog-agent/blob/666c775b3b0b408bc8ba4593f420fd526cd841a1/pkg/tagger/collectors/ecs_main.go#L118-L121](https://github.com/DataDog/datadog-agent/blob/666c775b3b0b408bc8ba4593f420fd526cd841a1/pkg/tagger/collectors/ecs_main.go#L118-L121)
6. Returns the resulting list of tags.

## Problems

1. The check whether the metadata endpoint in ECS exposes resource tags [https://github.com/DataDog/datadog-agent/blob/666c775b3b0b408bc8ba4593f420fd526cd841a1/pkg/util/ecs/detection.go#L83-L88](https://github.com/DataDog/datadog-agent/blob/666c775b3b0b408bc8ba4593f420fd526cd841a1/pkg/util/ecs/detection.go#L83-L88) can sometimes return `false`. This is possible just because it calls the `v3/<uuid>/taskWithTags` endpoint on the ECS agent which can fail with `Client.Timeout`. Note that at the moment the v1 and v3 client timeouts are 0.5 seconds and they are not configurable. This indeed sometimes happens in our case, especially on massive redeployments of all our ECS services or on datadog-agent restart. Apparently, the datadog agent tries to collect tags for all containers started/running on the host at once and this creates a sufficient load on the ECS agent (that serves the ECS metadata endpoints) to not being able to respond to each request in 0.5s. Note that the ECS agent in order to reply on `/taskWithTags` need to send at least 2 API calls to ECS API to get tags for the container instance and for the task, see: [https://github.com/aws/amazon-ecs-agent/blob/3776bee911c2267dedce46a9d2a17007f0bd73ae/agent/handlers/v2/response.go#L143-L163](https://github.com/aws/amazon-ecs-agent/blob/3776bee911c2267dedce46a9d2a17007f0bd73ae/agent/handlers/v2/response.go#L143-L163).
The logic of `HasEC2ResourceTags()` check is such that if it returned `false` at least once, the result is cached indefinitely: [https://github.com/DataDog/datadog-agent/blob/666c775b3b0b408bc8ba4593f420fd526cd841a1/pkg/util/ecs/detection.go#L138](https://github.com/DataDog/datadog-agent/blob/666c775b3b0b408bc8ba4593f420fd526cd841a1/pkg/util/ecs/detection.go#L138)
This means that after it fails once all metrics for all new containers on the host will be missing ECS tags because `addTagsForContainer` won't be called, see 4. This is obviously very bad.\
\
The fix prepared by the Datadog engineering team for us partially resolves this problem: [https://github.com/DataDog/datadog-agent/compare/7.23.x...7.23-ecs-ec2-tags-expiry](https://github.com/DataDog/datadog-agent/compare/7.23.x...7.23-ecs-ec2-tags-expiry). In case of failure the check result will be cached for 5 minutes meaning that all new containers will be missing ECS tags for only 5 minutes given that the next check in 5 minutes passes. This is better than it was before but still not ideal.

2. Another problem is item 5. If `addTagsForContainer` in 4. fails to get tags for a particular container from the ECS metadata endpoint (ECS agent running on the underlying host), e.g. due to client timeout, only the generic tags got in 3. are used and cached for the container. The agent will never retry to get ECS tags for that container again. We can see a lot of related logs confirming this behaviour:
    ```
    (pkg/tagger/collectors/ecs_main.go:124 in addTagsForContainer) | Unable to get resource tags for container 8bd7b34e144be4f1793e49914ad38066260bb0108af9341d70036e9eecc00dd5: failed to get task with tags from metadata v3 API: Get "[http://169.254.170.2/v3/645db862-98f2-4116-b342-564414e09994/taskWithTags](http://169.254.170.2/v3/645db862-98f2-4116-b342-564414e09994/taskWithTags)": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
    ```
    I think this behavior is wrong. If the user configured collection of resource logs by setting `DD_ECS_COLLECT_RESOURCE_TAGS_EC2`, the agent should collect them no matter what. If it fails it should retry later until succeeds.

## Workaround

I've implemented a few fixes/workarounds for both of these problems in my branch. You can see the changes here: [https://github.com/psviderski/datadog-agent/compare/7.23.x..pavel/7.23.1-ecs-tags-pavel-fix](https://github.com/psviderski/datadog-agent/compare/7.23.x..pavel/7.23.1-ecs-tags-pavel-fix). They are pretty simple:

1. Made the `HasEC2ResourceTags` check to always pass without calling `/taskWithTags` as I'm sure that the metadata endpoint in our ECS clusters exposes resource tags.
2. Increased the default timeouts for v1 and v3 ECS clients 0.5s → 5s.
3. Made the `addTagsForContainer` handler that collects tags for containers propagate an error if it fails to get tags from the ECS metadata endpoint. This error makes the `ECSCollector.Fetch` method to fail which forces the agent to not cache the incomplete tags for a container and retry tags collection later until it succeeds.

1 addresses the first problem. 2 and 3 address the second problem.

We have been running the datadog agent built from my branch for several days in our `dev` cluster and it seems to solve the issue with missing tags. We plan to test it in a bigger production cluster next week.

My change doesn't pretend to be merged to the upstream repo as is. I hope it can help you implement proper fixes for the problems I pointed out. Please let me know if you don't agree with something or need any additional information. I'm happy to raise a PR if you are happy with some of my changes or can tell me what to change.

### PR DESCRIPTION
**Pull request by juliogreff**
### What does this PR do?

Fixes an issue where the Agent would not retry resource tags collection for
containers on ECS if it could retrieve only a subset of tags. Now it will
keep on retrying until the complete set of tags is collected.

### Motivation

Fixes #6758

### Describe how to test your changes

I wasn't able to trigger the error condition on purpose (either all endpoints would fail, or none would). At the very least, we should test that the collection of ECS tags still works as expected.

### CODE REVIEW

File: pkg/tagger/local/tagstore.go [hunk 1]
```diff
@@ -92,6 +91,10 @@ func (s *tagStore) processTagInfo(tagInfos []*collectors.TagInfo) {
 			log.Tracef("processTagInfo err: empty source name, skipping message")
 			continue
 		}
+		if info.SkipCache {
```
Comment (ahmed-mez) [inline pkg/tagger/local/tagstore.go:94]: would it be useful to add some telemetry here?

**Review by ahmed-mez — APPROVED** [review]

File: pkg/tagger/local/tagstore.go (hunk 1 above)
Comment (juliogreff) [reply pkg/tagger/local/tagstore.go:94]: Added!

**Review by ogaca-dd — APPROVED** [review]
LGTM

**Comment by boh-dan** [comment]
@juliogreff I found an issue with your fix that leads to metrics sent without ECS tags. Can you please take a look https://github.com/DataDog/datadog-agent/issues/8341? I also have a PR that fixes this problem https://github.com/DataDog/datadog-agent/pull/8342.

**Pull request merged**